staging/src/k8s.io/cli-runtime/pkg/printers/kyaml.go

```diff
@@ -0,0 +1,463 @@
+	// We are using our own JSON here(instead of json.Unmarshal) because the
+	// Go JSON library doesn't try to pick the right number type (int, float,
+	// etc.) when unmarshalling to interface{}, it just picks float64
+	// universally.
+	var jsonObj interface{}
+	if err := utiljson.Unmarshal(jsonBytes, &jsonObj); err != nil {
+		return nil, fmt.Errorf("error unmarshaling from JSON: %w", err)
+	}
+	return jsonObj, nil
+}
+
+func (p *KYAMLPrinter) fromAny(obj any, w io.Writer) error {
+	buf := &strings.Builder{}
+	if err := p.formatValue(reflect.ValueOf(obj), 0, flagDefault, buf); err != nil {
+		return err
+	}
+	if _, err := fmt.Fprintf(w, "%s\n", buf.String()); err != nil {
+		return fmt.Errorf("error writing to output: %w", err)
+	}
+	return nil
+}
+
+type flagMask uint64
+
+const (
+	flagDefault   flagMask = 0
+	flagLazyQuote flagMask = 1 << iota
+)
+
+func (p *KYAMLPrinter) formatValue(v reflect.Value, indent int, flags flagMask, buf *strings.Builder) error {
+	if !v.IsValid() {
+		buf.WriteString("null")
+		return nil
+	}
+
+	// Handle things that implement json.Marshaler through that interface.
+	if v.CanInterface() {
+		if m, ok := v.Interface().(json.Marshaler); ok {
+			jsonBytes, err := m.MarshalJSON()
+			if err != nil {
+				return fmt.Errorf("error marshaling to JSON: %w", err)
+			}
+			jsonObj, err := p.unmarshalJSON(jsonBytes)
+			if err != nil {
+				return err
+			}
+			// If something chose to marshal as "null", we should respect that.
+			// Any nil-pointer will work.
+			if jsonObj == nil {
+				jsonObj = (*int)(nil)
+			}
+			v = reflect.ValueOf(jsonObj)
+		}
+	}
+
+	switch v.Kind() {
+	case reflect.String:
+		val := v.String()
+		if (flags&flagLazyQuote != 0) && !needsQuotes(val) {
+			fmt.Fprintf(buf, "%s", val)
+		} else {
+			fmt.Fprintf(buf, "%q", val)
+		}
+		return nil
+
+	case reflect.Bool:
+		val := v.Bool()
+		fmt.Fprintf(buf, "%v", val)
+		return nil
+
+	case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64:
+		val := v.Int()
+		fmt.Fprintf(buf, "%d", val)
+		return nil
+
+	case reflect.Uint, reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64:
+		val := v.Uint()
+		fmt.Fprintf(buf, "%d", val)
+		return nil
+
+	case reflect.Float32, reflect.Float64:
+		val := v.Float()
+		fmt.Fprintf(buf, "%g", val)
+		return nil
+
+	case reflect.Interface, reflect.Ptr:
+		if v.IsNil() {
+			buf.WriteString("null")
+			return nil
+		}
+		// Note: retain flags here.
+		return p.formatValue(v.Elem(), indent, flags, buf)
+
+	case reflect.Struct:
+		buf.WriteString("{\n")
+		if err := p.writeStructFields(v, indent, buf); err != nil {
+			return err
+		}
+		p.writeIndent(indent, buf)
+		buf.WriteString("}")
+		return nil
+
+	case reflect.Map:
+		if v.Len() == 0 {
+			buf.WriteString("{}")
+			return nil
+		}
+		buf.WriteString("{\n")
+
+		mapKeys := v.MapKeys()
+		sort.Slice(mapKeys, func(i, j int) bool {
+			return mapKeys[i].String() < mapKeys[j].String()
+		})
```

Comment:
If we're going to have our own YAML formatter, it would be cool if it could sort struct fields in the "right" order (ie, matching the golang type or the protobuf indexes). Or at least just ensure that the `TypeMeta` and `ObjectMeta` fields come before everything else? (ie, fix ConfigMaps to not sort `data` before `metadata.name`).

Reply:
I want this too, but kubectl doesn't always have a Go type for the thing it is unmarshalling.  It's not possible to do *in this code path*, without passing around some schema definition or else carrying something like managed fields just to list the sort order.  There's an open issue on sorting better, but I don't think we can conflate these topics.

Reply:
Maybe the sorting could be a prerequisite for enabling by default / moving out of alpha / something like that.